platform/linux-generic/pktio/loop.c

```diff
@@ -331,11 +339,21 @@ static int loopback_send(pktio_entry_t *pktio_entry, int index ODP_UNUSED,
 		 */
 		out_octets_tbl[i] = bytes;
 		nb_tx++;
+		pkt_hdr = packet_hdr(pkt_tbl[i]);
 
 		if (tx_ts_enabled && tx_ts_idx == 0) {
-			if (odp_unlikely(packet_hdr(pkt_tbl[i])->p.flags.ts_set))
+			if (odp_unlikely(pkt_hdr->p.flags.ts_set))
 				tx_ts_idx = i + 1;
 		}
+
+		if (odp_unlikely(tx_compl_enabled && pkt_hdr->p.flags.tx_compl &&
+				 pkt_hdr->tx_compl_mode == ODP_PACKET_TX_COMPL_ALL)) {
+			compl_info[i].tx_compl_q = pkt_hdr->tx_compl_queue;
+			compl_info[i].user_ptr = pkt_hdr->user_ptr;
+		} else {
+			compl_info[i].tx_compl_q = ODP_QUEUE_INVALID;
+			compl_info[i].user_ptr = NULL;
+		}
```

Comment:
Could these writes be avoided - so that compiler is happy and minimal amount of if statements / data initialization is added to the likely branch (not tx compl).

E.g. instead of compl_info[i].tx_compl_q signaling need to send event. There could be uint8_t send_compl[num] array of indexes that need tx compl. Another variable for number of packet/indexes (init to zero). In the common case number of indexes would remain zero, also when tx_compl_enabled.

Reply:
Should be more optimized in v2.

Reply:
When thinking this again. Could all/most tx compls be handled at the top level (in odp_pktout_send()). That would need extra information from driver telling if "sent" packets were freed or not, and modification to drivers generally not to free sent packets. Steps on top level would be (if tx compl enabled):
1. Record which packets request comp event (and their dest queue/ctx ptr)
2. When send returns, check if sent packets had requests: reuse sent packet if it's still available, or alloc new event if packet not available

Reply:
They certainly could. The downside is a slower still slowpath, if this is acceptable then sure.